numpy/core/_exceptions.py

```diff
@@ -122,20 +122,88 @@ class TooHardError(RuntimeError):
 
 @set_module('numpy')
 class AxisError(ValueError, IndexError):
-    """ Axis supplied was invalid. """
-    def __init__(self, axis, ndim=None, msg_prefix=None):
-        # single-argument form just delegates to base class
-        if ndim is None and msg_prefix is None:
-            msg = axis
+    """Axis supplied was invalid.
+
+    A `ValueError` and `IndexError` subclass raised whenever an
+    ``axis`` parameter is specified that is larger than
+    the number of array dimensions.
```

Comment:
```suggestion
    .. versionadded:: 1.13

    This is raised whenever an ``axis`` parameter is specified that is larger
    than the number of array dimensions.
    For compatibility with code written against older versions of numpy which
    raised a mixture of `ValueError` and `IndexError` for this situation, this
    exception subclasses both to ensure that ``except ValueError`` and
    ``except IndexError`` continue to catch `AxisError`s.
```

This is my fault for not writing a longer docstring in #8584 when I introduced this error.

Reply:
Done as of https://github.com/numpy/numpy/pull/19459/commits/7613ef481c437b45b315a464f85b4e52e42635d8.

I've moved around the `versionadded` directives a bit, moving the new `..versionadded:: 1.22` directive directly to the attributes section in the process, but I'm quite with content with how it looks of now.